File: main/lru-cache-lc/lru_cache_lc.rb
# A node for a circular doubly linked list.
class Node
  attr_reader :left, :right
  attr_accessor :key, :value

  def initialize(key = nil, value = nil)
    @left = @right = self

    @key = key
    @value = value
  end

  def splice_right(node)
    node.detach

    node.right = @right
    node.left = self

    @right = @right.left = node
  end

  protected

  attr_writer :left, :right

  def detach
    @left.right = @right
    @right.left = @left

    @left = @right = self
  end
end

NOT_FOUND = -1

# A cache of key-value mappings with a least-recently-used invalidation policy.
class LRUCache
  # :type capacity: Integer
  def initialize(capacity)
    raise ArgumentError, 'capacity must be positive' unless capacity.positive?
    @capacity = capacity
    @size = 0
    @list = Node.new # sentinel node
    @lookup = {} # maps keys to nodes
  end

  # :type key: Integer
  # :rtype: Integer
  def get(key)
    node = @lookup[key]
    return NOT_FOUND unless node
    bump(node)
    node.value
  end

  # :type key: Integer
  # :type value: Integer
  # :rtype: Void
  def put(key, value)
    node = @lookup[key]

    if node
      node.value = value
    elsif @size == @capacity
      node = @list.left
      @lookup.delete(node.key)

      node.key = key
      node.value = value

      @lookup[key] = node
    else
      raise 'Internal error: size exceeds capacity!' unless @size < @capacity

      @size += 1
      @lookup[key] = node = Node.new(key, value)
    end

    bump(node)
  end

  private def bump(node)
    @list.splice_right(node)
  end
end

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache.new(capacity)
# param_1 = obj.get(key)
# obj.put(key, value)

File: main/lru-cache-lc/lru-cache-lc.js
'use strict';

const NOT_FOUND = -1;

// A node in a circular doubly linked list.
function Node(prev, key, value) {
    this.key = key;
    this.value = value;

    if (prev) {
        // This is a non-sentinel node, inserted into an existing chain.
        this.attach(prev);
    } else {
        // This is a sentinel node.
        this.prev = this.next = this;
    }
}

Node.prototype.attach = function(prev) {
    this.prev = prev;
    this.next = prev.next;

    prev.next.prev = this;
    prev.next = this;
};

Node.prototype.detach = function() {
    this.prev.next = this.next;
    this.next.prev = this.prev;
};

/**
 * @param {number} capacity
 */
function LRUCache(capacity) {
    if (!Number.isInteger(capacity) || capacity <= 0) {
        throw new RangeError('capacity must be a positive integer');
    }

    this._capacity = capacity;
    this._length = 0;
    this._chain = new Node(); // Sentinel node.
    this._map = {};
}

/**
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    const node = this._map[key];

    if (node) {
        this._bump(node);
        return node.value;
    }

    return NOT_FOUND;
};

/**
 * @param {number} key
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    let node = this._map[key];

    if (node) {
        this._bump(node);
        node.value = value;
        return;
    }

    if (this._length < this._capacity) {
        ++this._length;
        node = new Node(this._chain, key, value);
    } else {
        node = this._chain.prev;
        this._bump(node);

        delete this._map[node.key];
        node.key = key;
        node.value = value;
    }

    this._map[key] = node;
};

LRUCache.prototype._bump = function(node) {
    node.detach();
    node.attach(this._chain);
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */